### ISSUE TITLE
document tests

### ISSUE THREAD

**Issue by petrelharp** [jguhlin/lostruct-py#1]
I'm having a look at the tests. Very nice! A few things would help do a better double-check the correctness/completeness:

- somewhere (the README?) it should say how to run the tests (install tox; run `tox`)

- I'm having trouble with actually running it - I've got python3.7, so at first tox complained that it couldn't find py38; so I edited `tox.ini` to be `py37`, and then it runs but says `No module named 'skbio'`, despite this being installed. I'm more familiar with the `unittest` framework, and would be happy to shift over these tests to that framework, if you don't have a strong opinion?

- I see the tests are mostly comparing to pre-calculated values; could you document where those values came from (ie, the code that was run to produce it)? E.g., in a comment next to the value, but we should also provide an R script to produce the values (especially, `lostruct-results/mds_coords.csv`).

**Comment by jguhlin** [comment]
It's actually py.test, I didn't have much luck getting tox to run successfully. Not sure why skbio becomes an issue as it's listed in the requirements.txt ?

Will throw in the documents. I ran lostruct using the command-line version but it should be in my history.

**Comment by petrelharp** [comment]
> It's actually py.test

Ah, ok. Well, the offer stands - I'm inclined to switch it to unittest since I'm familiar with it, and it would only take 5ish minutes. Want me to have a go?

**Comment by jguhlin** [comment]
Nah, let's stick with this. The offer is appreciated though! It's simple enough functions and already working with Travis CI.

**Comment by jguhlin** [comment]
Cool, added some more documentation about the tests, where the test data comes from, and did get tox to work, although I'm still installing multiple interpreters on my machine to make sure it all works. 

Let me know if that's sufficient, otherwise, we could generate an R script or something? The tests are now comparing the first PC as well (but squaring everything before looking at differences, since the sign likes to be opposite between R and Python....)

**Comment by petrelharp** [comment]
I'm working on getting the tests to run. `requirements.txt` isn't much help right now with my workflow because it has a whole bunch of modules that aren't used, and only with very specific versions? It looks like `requirements.txt` could just have
```
numpy
numba
cyvcf2
```
and a `requirements-dev.txt` could additionally have
```
scikit-bio
```
(since that's used for testing only).


> Nah, let's stick with this.

Ah, so you're pretty used to py.test then? Just to be clear: by not switching, you're putting an extra burden on me to learn an unfamiliar framework - which is fine, but I'd hope there's a good reason for it.

**Comment by petrelharp** [comment]
> added some more documentation about the tests

Great! You mean [here](https://github.com/jguhlin/lostruct-py/commit/6f8cd82a462bd5f4609168227660c2c9af99477f#diff-04c6e90faac2675aa89e2176d2eec7d8R66)? I think there's some context missing yet - I don't see any explanation in the tests themselves of where the numbers come from?  For instance, where's [this one](https://github.com/jguhlin/lostruct-py/blob/master/tests/test_fns.py#L49) come from?  Writing down these things now will save work in the long run...

**Comment by jguhlin** [comment]
Not used to it, I just looked to see what is most popular and easiest to use. unittest looks like it has extra boilerplate code, but if you want to throw some up as unittest, go ahead. Pytest does support running unittest's as well. So not a terribly good reason.

I created the requirements.txt as outlined in several places, I've tried reducing them down a bit but no luck. Are you still having issues running the tests? It's working with 3.7:
https://travis-ci.org/github/jguhlin/lostruct-py/jobs/720792788

I'll dig into the requirements.txt and see what I can remove without breaking things. Are you running things in a conda env or venv? Or installing globally? I'll see if I can create a conda yaml file as well.

I'll work on adding some more contexts to the tests! They are pretty weak right now anyways, it's basically, things look the same as the R version, so any numbers that are outputs I make sure they are within a tolerable error range. It'd be better to have something a bit more formal. But this does tell me if I screw something up right now (like I did when adding in sparse matrices originally).

**Comment by petrelharp** [comment]
> I'll dig into the requirements.txt and see what I can remove without breaking things. Are you running things in a conda env or venv? Or installing globally? I'll see if I can create a conda yaml file as well.

I'm installing things globally, perhaps embarassingly.  But - I think it should work with only the modules in my message above? (edited to add `sparse`) - these are the things that aren't core modules that get imported.

... and: the tests pass for me now!

And, btw: I believe these are good tests, and that they mean that things are probably right!

**Comment by jguhlin** [comment]
Yes, installing from pip it should install the dependencies, but then the tests will require scikit-bio, which has the pcoa function we use that is comparable to R's MDS. Perhaps it should be a requirement, even if none of the base code calls for it (at this stage) since it's an obvious analysis that will be done? Then pip should handle installing the dependencies required even for tests.

I'm turning into a container/venv/conda person. My typical work MO for python is to launch a conda env and launch jupyter / script from within it.

**Comment by petrelharp** [comment]
The usual way to do this is to have something like `requirements/conda-minimal.txt` and `requirements/development.txt`, so that if not running the tests you need the first list, and for the tests you need the second list.

**Comment by jguhlin** [comment]
Cool, I can work on that.

**Comment by jguhlin** [comment]
Although, if I add in the plotting 3 corners part, scikit bio becomes a dependency rather than a testing dependency, so maybe hold off?
